### PR TITLE
separate bmesh triangulate / tesselate

### PR DESCRIPTION
**Pull request by og76**

### CODE REVIEW

File: nodes/geometry/bmesh_tesselate.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+from ... events import propertyChanged
+from ... base_types.node import AnimationNode
+
+class BMeshTessellateNode(bpy.types.Node, AnimationNode):
+    bl_idname = "an_BMeshTessellateNode"
+    bl_label = "Triangulate BMesh"
+    bl_width_default = 160
+    searchLabels = ["Tesselate BMesh"]
+
+    quad = IntProperty(name = "Quad Method", update = propertyChanged, max = 3, min = 0)
```
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: I hate it that these methods are undocumented, but I also don't see a better way atm..

File: nodes/geometry/bmesh_tesselate.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    ngon = IntProperty(name = "Ngon Method", update = propertyChanged, max = 1, min = 0)
+
+    def create(self):
+        self.newInput("BMesh", "BMesh", "bm")
```
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: is the original bmesh object changed? If yes then 'dataIsModified' is needed here

File: nodes/geometry/bmesh_tesselate.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        self.newOutput("BMesh", "BMesh", "bm")
+        self.newOutput("Integer List", "Matching Ngon Indices", "ngonIndices", hide = True)
```
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: what is this used for? Do you have an example?
Is there maybe a better (more descriptiv) name?

File: nodes/geometry/bmesh_tesselate.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    def getExecutionCode(self):
+        isLinked = self.getLinkedOutputsDict()
+        if not any(isLinked.values()): return
+        
+        yield "faces = bm.faces[:]"
+        
+        # do not invert isLinked order, this needs original faces
+        if isLinked["ngonIndices"]: 
+            yield "ngonIndices = list( range(len(faces)) )"
```
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: just a little thing: these exra spaces aren't necessary in my opinion. Sometimes it is helpful to make code clearer this way, but I think this case here is not special enouph to break the (unwritten) rules.

File: nodes/geometry/bmesh_tesselate.py (hunk 1 above)
Comment (og76) [inline nodes/geometry/bmesh_tesselate.py]: they kinda switch the tris, but it's not clear to me how
I was wondering if I should keep them for advanced, I mean how often are they gonna be used?

File: nodes/geometry/bmesh_tesselate.py (hunk 2 above)
Comment (og76) [inline nodes/geometry/bmesh_tesselate.py]: not as far as I tested, 
I'll test again for more cases, loops etc

File: nodes/geometry/bmesh_tesselate.py (hunk 1 above)
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: yes, they should go into advanced

File: nodes/geometry/bmesh_tesselate.py (hunk 2 above)
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: If the input and output are different bmesh objects than the socket identifier has to be different as well, but I think the original data is modified. Will test as well

Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: bmesh copy
 dataIsModified has to be set to True.

File: nodes/geometry/bmesh_tesselate.py (hunk 3 above)
Comment (og76) [inline nodes/geometry/bmesh_tesselate.py]: Source ngon ..? Original ngon ..? Initial ngon indices?

it shows original poly that tri belongs to
like
source ngon index

can use this if you need to ref the original face, if you only use tria in a hidden way (in the end use full mesh) etc

Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: ah I see, thx

File: nodes/geometry/bmesh_tesselate.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    def drawAdvanced(self, layout):
+        layout.prop(self, "quad")
+        layout.prop(self, "ngon")
+
+    def getExecutionCode(self):
+        isLinked = self.getLinkedOutputsDict()
+        if not any(isLinked.values()): return
+        
+        yield "faces = bm.faces[:]"
```
Comment (JacquesLucke) [inline nodes/geometry/bmesh_tesselate.py]: last thing: is it important to make this copy?

**Comment by og76** [comment]
guess it works without copy, 
(never sure, I'm not that familiar with bmesh ways .., just get inspiration from here and there)
I don't remember, maybe I hoped it will keep intact after the triangulation, but not ..

**Comment by JacquesLucke** [comment]
ok.
yet another last thing: I think we should either use the term "tesselate" or "triangulate", not both. And I think "triangulate" is better because more people already know what it means.

**Comment by og76** [comment]
u mean no search tags?
and should I change the py file name also?

**Comment by JacquesLucke** [comment]
yes and yes.
for both files

**Comment by JacquesLucke** [comment]
thanks for your efforts. I think that was the last thing for v1.6

**Pull request merged**